stdlib/public/core/IntegerParsing.swift

```diff
@@ -2,23 +2,166 @@
 //
 // This source file is part of the Swift.org open source project
 //
-// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
 // Licensed under Apache License v2.0 with Runtime Library Exception
 //
 // See https://swift.org/LICENSE.txt for license information
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 //===----------------------------------------------------------------------===//
 
+@_alwaysEmitIntoClient
+internal func _parseASCIIDigits<Result: FixedWidthInteger>(
+  _ codeUnits: UnsafeBufferPointer<UInt8>, radix: Int, isNegative: Bool
+) -> Result? {
+  _internalInvariant(radix >= 2 && radix <= 36)
+  guard _fastPath(!codeUnits.isEmpty) else { return nil }
+  let numericalUpperBound, uppercaseUpperBound, lowercaseUpperBound: UInt8
+  if radix <= 10 {
+    numericalUpperBound = 48 /* "0" */ &+ UInt8(radix)
+    uppercaseUpperBound = 65
+    lowercaseUpperBound = 97
+  } else {
+    numericalUpperBound = 58
+    uppercaseUpperBound = 65 /* "A" */ &+ UInt8(radix &- 10)
+    lowercaseUpperBound = 97 /* "a" */ &+ UInt8(radix &- 10)
+  }
+  let multiplicand = Result(radix)
+  var result = 0 as Result
+  for digit in codeUnits {
+    let digitValue: Result
+    if _fastPath(digit >= 48 && digit < numericalUpperBound) {
+      digitValue = Result(digit &- 48)
+    } else if _fastPath(digit >= 65 && digit < uppercaseUpperBound) {
+      digitValue = Result(digit &- 65 &+ 10)
+    } else if _fastPath(digit >= 97 && digit < lowercaseUpperBound) {
+      digitValue = Result(digit &- 97 &+ 10)
+    } else {
+      return nil
+    }
+    let (temporary, overflow1) =
+      result.multipliedReportingOverflow(by: multiplicand)
+    guard _fastPath(!overflow1) else { return nil }
+    let (nextResult, overflow2) = isNegative
+      ? temporary.subtractingReportingOverflow(digitValue)
+      : temporary.addingReportingOverflow(digitValue)
+    guard _fastPath(!overflow2) else { return nil }
+    result = nextResult
+  }
+  return result
+}
+
+@_alwaysEmitIntoClient
+internal func _parseASCII<Result: FixedWidthInteger>(
+  _ codeUnits: UnsafeBufferPointer<UInt8>, radix: Int
+) -> Result? {
+  _internalInvariant(!codeUnits.isEmpty)
+  let first = codeUnits[0]
+  if first == 45 /* "-" */ {
+    return _parseASCIIDigits(
+      UnsafeBufferPointer(rebasing: codeUnits[1...]),
+      radix: radix, isNegative: true)
+  }
+  if first == 43 /* "+" */ {
+    return _parseASCIIDigits(
+      UnsafeBufferPointer(rebasing: codeUnits[1...]),
+      radix: radix, isNegative: false)
+  }
+  return _parseASCIIDigits(codeUnits, radix: radix, isNegative: false)
+}
+
+@_alwaysEmitIntoClient
+@inline(never)
+internal func _parseASCII<S: StringProtocol, Result: FixedWidthInteger>(
+  _ text: S, radix: Int
+) -> Result? {
+  var str = String(text)
```

Comment:
Do you know what code gets generated for this? Hopefully it will copy the UTF-8 contents for a Substring and bump a ref count for a String. I want to make sure it's not invoking some collection-of-`Character` initializer instead.

Reply:
Speaking of which, how is this code path being exercised and tested?

Reply:
You're right, this will need a test. What's the simplest way to write a test that bridges from an Obj-C string?

Reply:
As to the code that's generated, I took a look at a reduced example: https://godbolt.org/z/76qYrfebT

I'm not totally fluid with reading the output when it gets a little complicated, but when specialized for `Substring`, there is this line:
>         call    (lazy protocol witness table accessor for type Swift.Substring and conformance Swift.Substring : Swift.Sequence in Swift)

Reply:
Yeah, though for a bit-int cold path I don't care very much.

Reply:
The easiest way is probably to extend https://github.com/apple/swift/blob/main/test/stdlib/NSSlowString.swift with more test cases.